### PR TITLE
feat(mcp): structured output format for AI consumption

### LINKED ISSUE #3037: [feature] format API output for AI
check playwright mcp how they format output - we prob can improve ctx engineering here

### PR DESCRIPTION
**Pull request by yashksaini-coder**
## Summary

Closes #3037

Improves the MCP server's API output format for AI agent consumption, inspired by how playwright-mcp structures its output using `### Section` markdown headers that are machine-parseable.

## Problem

The previous output was raw JSON dumps and unstructured text, making it hard for AI agents to extract relevant information, understand what was returned, or know what to do next.

### New helper functions
| Helper | Purpose |
|---|---|
| `formatSection` | Consistent `### Title` markdown sections |
| `formatOptionalSection` | Conditionally rendered section |
| `formatQuery` | Echoes search parameters (like playwright's "Ran code" receipt) |
| `formatPagination` | Smart pagination hints with next-offset guidance |
| `formatDeepLink` | `screenpipe://frame/<id>` and timeline deep links |
| `formatToolError` | Categorized errors: network vs HTTP status vs validation |

### Tools enhanced
| Tool | Before | After |
|---|---|---|
| `health-check` | Raw JSON | `### Status / Capture stats / Warnings / Diagnostics` |
| `get-meeting` | Raw JSON dump | `### Meeting #N` metadata + `### Next steps` (transcript path) |
| `search-content` | Plain text list | `### Query` echo + deep links + `### Results / Next steps` |
| `activity-summary` | Unstructured text | `### Apps / Audio / Key content / Next steps` |
| `list-meetings` | Plain list | Duration, IDs, `### Next steps` for transcript access |
| `search-elements` | Header + flat list | `### Query` echo + frame deep links + `### Next steps` |
| `frame-context` | Flat line list | `### Frame context / URLs / Accessibility nodes / Next steps` |
| `keyword-search` | Raw results | `### Query` echo + frame deep links + `### Next steps` |
| catch block | Generic string | Categorized `### Error` via `formatToolError` |

## Key design decisions

- **Consistent `### Section` headers** — parseable by AI agents splitting on `### `
- **Query echo** — every search tool echoes back what parameters were used (so agents can verify their call was interpreted correctly)
- **Next steps** — exploratory tools suggest follow-up actions with exact parameter values to copy
- **Deep links** — `screenpipe://frame/<id>` per result for direct navigation
- **Uniform error format** — all errors (validation + caught) use `### Error` section

## Testing
- TypeScript compiles cleanly (`tsc --noEmit` passes with 0 errors)
- No functional logic changed — only output formatting

### CODE REVIEW

File: crates/screenpipe-connect/screenpipe-mcp/src/index.ts [hunk 1]
```diff
@@ -633,15 +714,25 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
...
+          q: args.q as string | undefined,
+          content_type: args.content_type as string | undefined,
+          start_time: args.start_time as string | undefined,
+          end_time: args.end_time as string | undefined,
+          app_name: args.app_name as string | undefined,
+          window_name: args.window_name as string | undefined,
+          speaker_name: args.speaker_name as string | undefined,
+          limit: args.limit as number | undefined,
+          offset: args.offset as number | undefined,
+        });
```
[the remark this answers is not in the document]
Comment (yashksaini-coder) [reply crates/screenpipe-connect/screenpipe-mcp/src/index.ts]: Fixed in 8ddfbea — `search-content` now passes all `args` directly to `formatQuery` instead of a hand-picked subset, so every provided param (including `speaker_ids`, `min_length`, `max_length`, `include_frames`, `max_content_length`, etc.) is echoed back.

File: crates/screenpipe-connect/screenpipe-mcp/src/index.ts [hunk 2]
```diff
@@ -853,7 +974,7 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
             content: [
               {
                 type: "text",
-                text: "No elements found. Try: broader search, different role/source, or wider time range.",
+                text: "### Result\nNo elements found.\n\n### Next steps\n- Broaden search: try a different role, source, or wider time range\n- Use `frame-context` with a specific frame_id for full page structure",
               },
```
[the remark this answers is not in the document]
Comment (yashksaini-coder) [reply crates/screenpipe-connect/screenpipe-mcp/src/index.ts]: Fixed in 8ddfbea — the empty-result branch in `search-elements` now prepends `formatQuery(args)` so the query echo is present even when no results are found, consistent with all other handlers.

File: crates/screenpipe-connect/screenpipe-mcp/src/index.ts [hunk 3]
```diff
@@ -1308,21 +1548,39 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
...
         const results = data.data || [];
         if (results.length === 0) {
-          return { content: [{ type: "text", text: "No keyword search results found." }] };
+          return {
+            content: [
+              {
+                type: "text",
+                text: `### Query\n- query: ${args.query || "(none)"}\n\n### Result\nNo keyword search results found.\n\n### Next steps\n- Try \`search-content\` for semantic search across OCR and audio content`,
+              },
+            ],
```
[the remark this answers is not in the document]
Comment (yashksaini-coder) [reply crates/screenpipe-connect/screenpipe-mcp/src/index.ts]: Fixed in 8ddfbea — replaced the hardcoded template string with `formatQuery(args)` + `formatSection` helpers. All provided args (limit, offset, time range, etc.) are now echoed, not just `query`.

File: crates/screenpipe-connect/screenpipe-mcp/src/index.ts [hunk 4]
```diff
@@ -1133,10 +1280,73 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
...
+          monitors?: string[];
+          pipeline?: Record<string, unknown>;
+          audio_pipeline?: Record<string, unknown>;
+          pool_stats?: Record<string, unknown>;
         };
+
+        const now = Date.now();
+        const ageStr = (ts?: string) => {
+          if (!ts) return "never";
+          const diff = Math.round((now - new Date(ts).getTime()) / 1000);
```
[the remark this answers is not in the document]
Comment (yashksaini-coder) [reply crates/screenpipe-connect/screenpipe-mcp/src/index.ts]: Fixed in 8ddfbea — `ageStr` now guards with `Number.isNaN(date.getTime())` and returns `"invalid date"` for unparseable timestamps, and `"in the future"` for negative diffs instead of emitting `-30s ago`.

File: crates/screenpipe-connect/screenpipe-mcp/src/index.ts [hunk 5]
```diff
@@ -605,6 +605,87 @@ async function fetchAPI(
...
+ * when only a page limit was used.
+ */
+function formatPagination(
+  count: number,
+  total: number | null | undefined,
+  offset: number,
+  toolHint: string
+): string {
+  if (total != null && count < total) {
+    return `> ${count} of ${total} results — use \`${toolHint}\` with \`offset=${offset + count}\` for more`;
```
[the remark this answers is not in the document]
Comment (yashksaini-coder) [reply crates/screenpipe-connect/screenpipe-mcp/src/index.ts]: Fixed in 8ddfbea — condition changed from `count < total` to `offset + count < total`, and the hint now displays the shown range as `X–Y of Z` so the offset math is correct on any page.

File: crates/screenpipe-connect/screenpipe-mcp/src/index.ts [hunk 6]
```diff
@@ -605,6 +605,87 @@ async function fetchAPI(
...
+  if (httpMatch) {
+    const status = httpMatch[1];
+    const detail =
+      status === "401" ? "Invalid or missing API key."
+      : status === "403" ? "Access denied."
+      : status === "404" ? "Resource not found."
+      : status === "503" ? "screenpipe service is unavailable."
+      : `Server returned ${status}.`;
+    return `### Error\n${detail} (${toolName})`;
+  }
```
[the remark this answers is not in the document]
Comment (yashksaini-coder) [reply crates/screenpipe-connect/screenpipe-mcp/src/index.ts]: Fixed in 8ddfbea — broadened the regex to `/HTTP(?:\s+error)?:?\s+(\d+)/i` which matches both `HTTP error: 404` and `HTTP 404` forms, so `export-video`'s thrown messages are now correctly categorized.

**Comment by louis030195** [comment]
Thanks for the rework — the structured output direction is right, and you addressed all of Copilot's notes cleanly in `8ddfbea`. A few things still worth tightening before this lands:

**Deep-link inconsistency.** `formatDeepLink` emits markdown `[frame N](screenpipe://frame/N)`, but three call sites bypass it and build raw `screenpipe://frame/${id}` strings inline:
- `search-elements` (~line 991)
- `keyword-search` (~line 1561)
- `frame-context` metaLines (~line 1393)

Route all four through `formatDeepLink` so a future change to the URL scheme is one edit, not four.

**`health-check` Diagnostics section still dumps raw JSON.** `JSON.stringify(data.pool_stats)`, `data.pipeline`, `data.audio_pipeline` go in unformatted — the noisiest part of the response is exactly where structure would help most. Either flatten these into key/value lines, or drop them from the default response and add a `verbose: true` arg.

**Type cast lies.** `data.windows as Array<{ browser_url: string }>` should be `browser_url?: string` — the field is optional upstream. The repeated `args as Record<string, string | number | boolean | null | undefined>` casts are also unchecked; `formatQuery` happens to handle weird shapes safely but the cast itself is fragile. Either narrow at the entry point or use `unknown`.

**`get-meeting` `detection_source || \"auto\"`.** `auto` isn't a real detection source — it'll mislead agents that try to filter on it. Use `\"unknown\"` to match how other fields fall back.

**No tests.** `formatPagination` has off-by-one math that Copilot already caught once — that and `formatToolError`'s error categorization are exactly the kind of pure functions that deserve unit tests. \"TypeScript compiles cleanly\" is a build, not a test.

**CI is BLOCKED with no checks reporting on the branch** — worth checking why before relying on green-light.

None of this is merge-blocking; happy to land it as-is and follow up, or take another pass to clean these up. Your call.

**Comment by yashksaini-coder** [comment]
Thanks for the thorough review @louis030195! All five issues are addressed in commit 34bac35f6:

1. **Raw deep-link bypass** — all three call sites (`search-elements`, `frame-context`, `keyword-search`) now route through `formatDeepLink()` instead of building raw URL strings.

2. **Health-check Diagnostics JSON** — `pool_stats`, `pipeline`, `audio_pipeline` now use the new `flattenObject()` helper which renders one `key: value` line per field instead of a raw JSON blob.

3. **Type casts** — `formatQuery` signature updated to accept `Record<string, unknown>` and filter for primitive values internally, eliminating all `args as Record<string, string | number | ...>` casts. `browser_url` changed from `string` to `browser_url?: string`.

4. **`detection_source || "auto"`** — changed to `"unknown"` (the correct sentinel for a missing detection source).

5. **Unit tests** — all formatting helpers extracted to `src/helpers.ts` (exported) and covered by 28 test cases in `src/helpers.test.ts`, including: `formatPagination` offset math and boundary conditions, `formatToolError` HTTP regex patterns and ECONNREFUSED, `formatDeepLink` frameId=0 edge-case, `formatQuery` non-primitive filtering, and `flattenObject`.

All 40 tests pass (`bun test`), TypeScript compiles clean (`tsc --noEmit`).

**Comment by yashksaini-coder** [comment]
@louis030195 Hi I rebased and resolved git merge conflicts please review it now

**Comment by louis030195** [comment]
Closing — detected AI slop.

This is a 568-line "make output AI-consumable" markdown-formatting refactor (formatSection / formatPagination / formatDeepLink / formatToolError helpers across every MCP tool) without a concrete bug or failing agent. Broad output-shape refactors with no specific failure case are a common LLM-generated PR pattern.

If a specific MCP client/agent is choking on a specific tool's output, please open an issue with the agent + tool + actual failure mode first. We can then evaluate the right shape — which may not be markdown sections.
